core/services/health/health.go

```diff
@@ -0,0 +1,212 @@
+package health
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+// Types requiring health checks should implement the Checkable interface.
+type Checkable interface {
+	// Checkables should return nil if ready, or an error message otherwise.
+	Ready() error
+	// Checkables should return nil if healthy, or an error message otherwise.
+	Healthy() error
+}
+
+//go:generate mockery --name Checker --output ../../internal/mocks/ --case=underscore
+type (
+	// Checker provides a service which can be probed for system health.
+	Checker interface {
+		// Register a service for health checks.
+		Register(name string, service Checkable) error
+		// Unregister a service.
+		Unregister(name string) error
+		// IsReady returns the current readiness of the system.
+		// A system is considered ready if all checks are passing (no errors)
+		IsReady() (ready bool, errors map[string]error)
+		// IsHealthy returns the current health of the system.
+		// A system is considered healthy if all checks are passing (no errors)
+		IsHealthy() (healthy bool, errors map[string]error)
+
+		Start() error
+		Close() error
+	}
+
+	checker struct {
+		srvMutex   sync.RWMutex
+		services   map[string]Checkable
+		stateMutex sync.RWMutex
+		state      map[string]State
+
+		stop chan struct{}
+		done chan struct{}
+
+		utils.StartStopOnce
+	}
+
+	State struct {
+		ready   error
+		healthy error
+	}
+
+	Status string
+)
+
+var _ Checker = (*checker)(nil)
+
+const (
+	StatusPassing Status = "passing"
+	StatusFailing Status = "failing"
+
+	interval = 15 * time.Second
+)
+
+var (
+	metrics = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "health",
+		Help: "Health status by service",
+	},
+		[]string{"service_id"},
+	)
+)
+
+func NewChecker() Checker {
+	c := &checker{
+		services: make(map[string]Checkable, 10),
+		state:    make(map[string]State, 10),
+		stop:     make(chan struct{}),
+		done:     make(chan struct{}),
+	}
+
+	return c
+}
+
+func (c *checker) Start() error {
+	return c.StartOnce("HealthCheck", func() error {
+		// update immediately
+		c.update()
+
+		go c.run()
+
+		return nil
+	})
+}
+
+func (c *checker) Close() error {
+	return c.StopOnce("HealthCheck", func() error {
+		close(c.stop)
+		<-c.done
+		return nil
+	})
+}
+
+func (c *checker) run() {
+	defer close(c.done)
+
+	ticker := time.NewTicker(interval)
+
+	for {
+		select {
+		case <-ticker.C:
+			c.update()
+		case <-c.stop:
+			return
+		}
+	}
+
+}
+
+func (c *checker) update() {
+	state := make(map[string]State, len(c.services))
+
+	c.srvMutex.RLock()
+
+	for name, s := range c.services {
+		ready := s.Ready()
+		healthy := s.Healthy()
+
+		state[name] = State{ready, healthy}
+	}
+
+	c.srvMutex.RUnlock()
```

Comment:
hmm not sure if worth it, but you could copy the map inside the mutex, to have it locked for a shorter time

Reply:
Good point, it depends on how intensive the checks are. I think this might be one of few uses for `sync.Map`

> (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow,  [...]